packages/toolkit/core-kit/src/password-policy.ts

```diff
@@ -0,0 +1,317 @@
+    words: Word[];
+  };
+};
+
+/** Password policy configuration guard. */
+const passwordPolicyGuard: z.ZodType<PasswordPolicy> = z.object({
+  length: z.object({
+    min: z.number().int().min(1),
+    max: z.number().int().min(1),
+  }),
+  characterTypes: z.object({
+    min: z.number().int().min(1).max(4),
+  }),
+  rejects: z.object({
+    pwned: z.boolean(),
+    repetitionAndSequence: z.boolean(),
+    words: z.array(z.object({ type: z.enum(['custom', 'personal']), value: z.string() })),
+  }),
+});
+
+/** The code of why a password is rejected. */
+type PasswordRejectionCode =
+  | 'too_short'
+  | 'too_long'
+  | 'character_types'
+  | 'unsupported_characters'
+  | 'pwned'
+  | 'repetition'
+  | 'sequence'
+  | 'restricted_words';
+
+/** A password issue that does not meet the policy. */
+type PasswordIssue = {
+  /** Issue code. */
+  code: `password_rejected.${PasswordRejectionCode}`;
+  /** Interpolation data for the issue message. */
+  interpolation?: Record<string, unknown>;
+};
+
+/**
+ * The class for checking if a password meets the policy. The policy is defined as
+ * {@link PasswordPolicy}.
+ *
+ * @example
+ * ```ts
+ * const checker = new PasswordPolicyChecker({
+ *   length: { min: 8, max: 256 },
+ *   characterTypes: { min: 2 },
+ *   rejects: { pwned: true, repetitionAndSequence: true, words: [] },
+ * });
+ *
+ * const issues = await checker.check('123456');
+ * console.log(issues);
+ * // [
+ * //   { code: 'password_rejected.too_short' },
+ * //   { code: 'password_rejected.character_types', interpolation: { min: 2 } },
+ * //   { code: 'password_rejected.pwned' },
+ * //   { code: 'password_rejected.sequence' },
+ * // ]
+ * ```
+ */
+export class PasswordPolicyChecker {
+  static symbols = Object.freeze('!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~' as const);
+
+  constructor(public readonly policy: PasswordPolicy) {
+    // Validate policy.
+    passwordPolicyGuard.parse(policy);
+  }
+
+  /**
+   * Check if a password meets the policy.
+   *
+   * @param password - Password to check.
+   * @returns An array of issues. If the password meets the policy, an empty array will be returned.
+   */
+  /* eslint-disable @silverhand/fp/no-mutating-methods */
+  async check(password: string): Promise<PasswordIssue[]> {
+    const issues: PasswordIssue[] = [];
+
+    if (password.length < this.policy.length.min) {
+      issues.push({
+        code: 'password_rejected.too_short',
+      });
+    } else if (password.length > this.policy.length.max) {
+      issues.push({
+        code: 'password_rejected.too_long',
+      });
+    }
+
+    const characterTypes = this.checkCharTypes(password);
+    if (characterTypes === 'unsupported') {
+      issues.push({
+        code: 'password_rejected.unsupported_characters',
+      });
+    } else if (!characterTypes) {
+      issues.push({
+        code: 'password_rejected.character_types',
+        interpolation: { min: this.policy.characterTypes.min },
```

Comment:
nit: should we also tell the number of current char types?

Reply:
per design no need i think